### ISSUE TITLE
Error while sending Transaction

### ISSUE THREAD

**Issue by Cyrois** [horizontalsystems/bitcoin-kit-android#490]
Some Background Information: 

We are trying to use your open source library to build an electrumX backed wallet. So far we were able to replace the MerkleBlock code with regular BlockHeaders. Our BlockSyncer refactor was complete. We then replaced the receive Transaction code to use ElectrumX's protocol methods. 

We forked your library from checksum: `56758ce3eb85508c80ae3713bced9e260d0b3b1f`

Problem:
We are not experiencing issues when broadcasting a transaction using the original library code.

When we first tried to send I generated the following hash:

`020000000001013127276691d2e0ecc36957ad02e485c96bf8544083468ca7f39fa29a04556b1400000000000000000002ff1d0000000000001976a91487f9d504bae645c1851c1a91d1f4da06c8c02d2788acce810000000000001976a914f313c99918f399bbcab2bc34252a705a6ed56e1d88ac024830450221008682bf10945ef7ff53a66209382ce37e11a9adbee1fa0ff79e994ef6db888a2902202d7e05bbe44ed40168197b99338427f59c56254010e03e2f7be33dc012f003cc0121024383a3c606458892464b786a6910c8950f01c02df5be3511cb8e668d14704b1a00000000`

When I try to broadcast this Txn I get an error:
`Error validating transaction: Transaction 886187c8b7c35fec701940bbefb2d4ed657b14d353b8d1ee338824e1841c7119 orphaned, missing reference 146b55049aa29ff3a78c46834054f86bc985e402ad5769c3ece0d29166272731.`

We found that this was an issue when spending a segwit transaction. We had to modify the transactionCreator to use the txid instead of the txhash when spending a segwit txn. 

`
// add inputs
        for (output in selectedOutputsInfo.outputs) {
            val previousTx = checkNotNull(output.transaction) {
                throw BuilderException.NoPreviousTransaction()
            }
            val txInput = TransactionInput().apply {
                if(previousTx.segwit) {
                    val txIdBytes = HashUtils.toBytesAsLE(previousTx.txId)
                    previousOutputHash = txIdBytes
                    previousOutputHexReversed = previousTx.txId
                } else {
                    previousOutputHash = previousTx.hash
                    previousOutputHexReversed = previousTx.hashHexReversed
                }
                previousOutputIndex = output.index.toLong()
            }
            txInput.previousOutput = output
            transaction.inputs.add(txInput)
        }
`

With these changes we are able to generate a transaction hash...

`01000000000101eb6446b4ebdeac9bf21c0c19e543abc8fe3f998ce1c29d2b7d1e27e4225fed4200000000000000000002ff1d0000000000001976a91487f9d504bae645c1851c1a91d1f4da06c8c02d2788accc7d0000000000001976a914ef7c6c937ea03233b128978498d8566c5c9f7a5688ac02483045022100a56272ee264d47d6d5505c152edf27c8c1b411580f2c9309156f9f012e3de4e5022059dc8665b623c171450bb0b96fd68401a134c1a3714836409c04665de1e26daf012103eaf259596ed97fe4f74974fed4133562ede3b29326c191e274449ef77b2e4e4b00000000`

We get the following error 

`Error validating transaction: Error running script for input 0 referencing 42ed5f22e4271e7d2b9dc2e18c993ffec8ab43e5190c1cf29bacdeebb44664eb at 0: Script was NOT verified successfully..`

Notes:
1. The is an issue generating a segwit address. We had to modify the transactionBuilder to use the txid instead of the txhash when spending a segwit txn. 

2. When we receive a transaction, we are getting a hex value from ElectrumX, creating a BitcoinInput object and sending it through your parser.

This is the hex value that ElectrumX gives us...
`0100000000010130d33ae221d68c40bb70ff5097bc64643a5cc5556ab6539d514831a2618f05d10000000000ffffffff023b1e4a0000000000160014388711254d485e781b2c99bfdeba9cc72eed8a8f5a7101000000000016001428f4e589c78aae4efaf217a18dde8524bfe2eb2d024730440220354fa47694afa53407fe7a7a2a32458320fa27e9fb7976ed9942a22ce50e4ccf022047c00c3d9351119cacd954de1efc482470e12f9c01df7bda8f4fffb1e3e6fd8701210369b1152bcad103eaea7257ea6754836d7f8767df7f42582b3b12354f83a6803f00000000`

3. We confirmed that Segwit Transactions are indeed working on the latest master branch of the bitcoin-kit-android.

Comparing the hex that the master branch generated with ours, we found 2 differences. The version was different and the locktime was different. 

For now, we do not feel that we should need to modify your code to make this work. Running out of ideas why this is not working...

**Comment by Cyrois** [comment]
Using the branch that we forked from, it looks like spending a segwit transaction results in a loss...

`010000000126a6ce2bd979f3a9524b2630df8dcdb7f46782d4eac07b314cfa46cb6c1a8590010000006a47304402201630db7c9141c6cd4cafe6f96765d857ebb399b60eda73ddecd91092c50958c102206c51502ee5db29be4b63f1f265a17cedc8b665f35d87345119b3967da70728d401210284e258c7eadc92bca50199c9a35841b43b4537ec969c46f0f468b0ce303d5988000000000210270000000000001600143acf57afac22b3c9920f4bb72cb2447449e3026cfd272d01000000001976a914796d290757049590819dad0aca516b8eccbd97c588ac00000000`

Gives me an error of 
`
Error validating transaction: Transaction 90851a6ccb46fa4c317bc0ead48267f4b7cd8ddf30264b52a9f379d92bcea626 referenced by input 0 of f79cde0447e9aa7ba0ffd91b607f1f8602b541b7503809714d070297dfd393ac has already been spent..
`

**Comment by abdrasulov** [comment]
Hi @Cyrois,

I'd suggest finding the difference between building transactions in your forked version and the master and apply it to your code. There are steps in building the transaction. You can see these steps in `TransactionBuilder::buildTransaction()`. Debugging every step may help to find the difference.

As for the error "... has already been spent", the output you're trying to send has already been spent in another transaction.

Since the issue is not reproducible in the current version, we close it.

**Comment by abdrasulov** [comment]
@Cyrois,

I've checked the version you've forked. I reproduced the issue in that version. Unfortunately, there is no quick fix. It may take about a week. 

As you've noticed, we switched our database from Realm to SQLite since then. Also, there have been other fixes there. 

Summarizing the above, I would recommend you to update your code to the latest master version. And keep it up to date. It may help with applying new features and fixing potential issues in the future.
